libvast/src/arrow_table_slice.cpp

```diff
@@ -167,68 +200,71 @@ void decode(const type& t, const arrow::Array& arr, F& f) {
+    }
+    case arrow::Type::STRUCT: {
+      return decode(t, dynamic_cast<const arrow::StructArray&>(arr), f);
     }
     // -- lift floating point values to real -----------------------------
     case arrow::Type::HALF_FLOAT: {
       using array_type = arrow::NumericArray<arrow::HalfFloatType>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::FLOAT: {
       using array_type = arrow::NumericArray<arrow::FloatType>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::DOUBLE: {
       using array_type = arrow::NumericArray<arrow::DoubleType>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     // -- lift singed values to integer ----------------------------------
     case arrow::Type::INT8: {
       using array_type = arrow::NumericArray<arrow::Int8Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::INT16: {
       using array_type = arrow::NumericArray<arrow::Int16Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::INT32: {
       using array_type = arrow::NumericArray<arrow::Int32Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::INT64: {
       using array_type = arrow::NumericArray<arrow::Int64Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     // -- lift unsinged values to count ----------------------------------
     case arrow::Type::UINT8: {
       using array_type = arrow::NumericArray<arrow::UInt8Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::UINT16: {
       using array_type = arrow::NumericArray<arrow::UInt16Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::UINT32: {
       using array_type = arrow::NumericArray<arrow::UInt32Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
     }
     case arrow::Type::UINT64: {
       using array_type = arrow::NumericArray<arrow::UInt64Type>;
-      return decode(t, static_cast<const array_type&>(arr), f);
+      return decode(t, dynamic_cast<const array_type&>(arr), f);
```

Comment:
Revert to `static_cast` and change clang-tidy rules accordingly.

Reply:
Yes please. If the type is known you should use `static_cast`.